src/index.js

```diff
@@ -1,189 +1,222 @@
+                    directDomains: directDomains.join(',')
+                }
+            };
+        }
+
+        this.sauceStorage = new SauceStorage(this.username, this.accessKey);
+
+        wd.configureHttp({
+            retryDelay: WEB_DRIVER_CONFIGURATION_RETRY_DELAY,
+            retries:    WEB_DRIVER_CONFIGURATION_RETRIES,
+            timeout:    WEB_DRIVER_CONFIGURATION_TIMEOUT
+        });
+
+    }
+
+    _log (message) {
+        if (this.options.connectorLogging)
+            process.stdout.write(message + '\n');
+    }
+
+    async _getFreeMachineCount () {
+        var params = {
+            method: 'GET',
+            url:    [`https://${SAUCE_API_HOST}/rest/v1/users`, this.username, 'concurrency'].join('/'),
+            auth:   { user: this.username, pass: this.accessKey }
+        };
+
+        var response = await requestPromised(params);
+
+        return JSON.parse(response.body).concurrency[this.username].remaining.overall;
+    }
+
+    async getSessionUrl (browser) {
+        var sessionId = await browser.getSessionId();
+
+        return `https://app.${SAUCE_API_HOST}/tests/${sessionId}`;
+    }
+
+    async startBrowser (browser, url, jobOptions = {}, timeout = null) {
+        jobOptions = { ...jobOptions, ...browser };
+
+        var webDriver = wd.promiseChainRemote(`ondemand.${SAUCE_API_HOST}`, 80, this.username, this.accessKey);
+
+        var pingWebDriver = () => webDriver.eval('');
+
+        webDriver.once('status', () => {
+            // HACK: if the webDriver doesn't get any command within 1000s, it fails
+            // with the timeout error. We should send any command to avoid this.
+            webDriver.pingIntervalId = setInterval(pingWebDriver, WEB_DRIVER_PING_INTERVAL);
+
+            if (this.options.connectorLogging) {
+                this
+                    .getSessionUrl(webDriver)
+                    .then(sessionUrl => this._log(`${browser.browserName} started. See ${sessionUrl}`));
+            }
+        });
+
+        const {
+            idleTimeout = WEB_DRIVER_IDLE_TIMEOUT,
+
+            name             = jobOptions.jobName,
+            tunnelIdentifier = this.tunnelIdentifier,
+
+            ...additionalOptions
+        } = jobOptions;
+
+        var initParams = {
+            ...additionalOptions,
+
+            idleTimeout,
+
+            ...name && {
+                name
+            },
+
+            ...tunnelIdentifier && {
+                tunnelIdentifier
+            },
+
+            ...timeout && {
+                maxDuration: timeout
+            }
+        };
+
+        // NOTE: If IE11 is used, the "Display intranet sites in Compatibility View" option should be disabled.
+        // We do this this via the 'prerun' parameter, which should run our script on the Sauce Labs side,
+        // before the browser starts.
+        if (browser.browserName.toLowerCase() === 'internet explorer' && /11(\.\d*)?$/.test(browser.version)) {
```

Comment:
[minor/recommendation]
It'd be nice to keep in constants (except test files of course):
1. all regular expression,
2. all number except `-1,0,1`,
3. all strings and templates, if it's possible

Also I'm sure there's some eslint plugin which may help.